### ISSUE TITLE
Add command line options to manipulate imports

### ISSUE THREAD

**Issue by yatli** [Microsoft/msbuild#3691]
msbuild implicitly import files during various stages. there is no way to inspect or modify what is imported, at the moment. 

for example (https://github.com/dotnet/sdk/issues/887), sometimes we find overriding a property with /p flag is too late, but the only chance of early import relies on heuristics in finding `Directory.Build.props` -- a wildcard that might be "too greedy" (https://github.com/Microsoft/msbuild/issues/762).

can we get command line options to manipulate such behaviors? e.g. if specified, overrides the above heuristic and use a given file instead, or list the files imported, etc.

**Comment by dasMulli** [comment]
In theory, `-p:DirectoryBuildPropsPath=/path/to/some.props` should already work.

Would be interesting to change `_DirectoryBuildPropsFile` to a public property to allow changing the file name. 
Currently `-p:_DirectoryBuildPropsFile=foo.props` should already work and change the file name from `Directory.Build.props` to `foo.props` while maintaining the hierarchy search logic.
Is that what you are looking for?

**Comment by dasMulli** [comment]
There's also a `ImportDirectoryBuildProps` one could set to `false`

**Comment by dasMulli** [comment]
No, `DirectoryBuildPropsPath` is the full path, `_DirectoryBuildPropsFile` would do the lookup but it's not really a public property (hence the underscore).
It would be interesting if MSBuild maintainers would be accepting a change to make this property public (remove the underscore). 
Personally, I'd rather not and require users to always use `Directory.Build.props` for consistency. In there, build authors can add additional logic to search for other files or import other known files.

`/p`(`-p`) sets global properties which are present from the beginning and can't be overwritten by the static portion of project files (with an exception involving `<Project TreatAsLocalProperty="…">`) so you don't need to worry about it applying too late.

**Comment by yatli** [comment]
overriding `DirectoryBuildPropsPath` works for me, thanks @dasMulli !

I too would rather not let users create arbitrary recursive lookups. `Directory.Build.props` at least looks like a build system-related file, while arbitrary `.props` file may be everywhere in the filesystem.

I'm still not very sure about `/p` flag (also discussed here: https://github.com/Microsoft/msbuild/issues/1603)

If I specify `/p:BaseIntermediateOutputPath=` at command line, I get inconsistent result -- the intermediate path indeed moved to the desired location, but `project.json` etc. remained in the `obj` folder, hence my concern about "too late". :)

**Comment by dasMulli** [comment]
The `BaseIntermediateOutputPath` trouble isn't/wasn't related to the order of definitions but rather NuGet and MSBuild using different properties for essentially the same concerns.
If you still have troubles with it, I'd suggest logging a fresh issue somewhere to discuss implications. The most difficult part about it is making sure that restore runs and builds have the same global properties set via CLI parameters. I believe `Directory.Build.props` may be the best option here to ensure a consistent experience. 

> Directory.Build.props at least looks like a build system-related file

It is also more in-line with `Directory.Build.rsp` which is not implemented in MSBuild files (common props/targets) but in MSBuild itself, thus the name can't really be changed in a "good/consistent/compatible" way.

**Comment by dasMulli** [comment]
ad `BaseIntermediateOutputPath`: also see https://github.com/Microsoft/msbuild/issues/3695.
TL;DR don't set it for solutions.
But I think it should work for stand-alone projects. Or not using a rooted property / common location.

**Comment by yatli** [comment]
I think this issue can be closed now -- recommend documenting `DirectoryBuildPropsPath` somewhere though. Thanks @dasMulli for clarifying everything!

**Comment by yatli** [comment]
btw I'm rolling a cmake-dotnet integration module, FYI:

https://github.com/Z3Prover/z3/issues/1423

https://github.com/Microsoft/GraphEngine/blob/master/cmake/FindDotnet.cmake

**Comment by yatli** [comment]
@dasMulli some new findings here, even if I override `DirectoryBuildPropsPath`, during package restore, `Directory.Import.props` is choosen and my custom props file is ignored.

This brings complexity to parameterized package reference, e.g. `<PackageReference Id="..." Version="$(SOME_DEP_VERSION)" />` -- as our package version is refreshed on build, we have to update the root `Directory.Import.props` also on build -- and this breaks the out-of-tree building principle.

**Comment by dasMulli** [comment]
@yatli that sounds more like a caching issue.. do you use `-restore` to restore and build during the build or `-t:Restore;Build`? (which doesn't work if you change xml content in between)
Do you have a sample project that reproduces this behaviour?

**Comment by yatli** [comment]
@dasMulli I use `dotnet restore /p:....`.

looks like `DirectoryBuildPropsPath` expects a full path, and it correctly picks up the custom props when I give it a proper path.

**Comment by dasMulli** [comment]
yeah if it isn't a full path, it will be relative to the project directory. And if you do it on a solution, it will be relative to each project individually..
